src/parse/asp/exec.go

```diff
@@ -96,6 +185,167 @@ func execGetCachedOutput(key execKey, args []string) (output string, found bool)
 	}
 }
 
+// execGitBranch returns the output of a git_branch() command.
+//
+// git_branch() returns the output of `git symbolic-ref -q --short HEAD`
+func execGitBranch(s *scope, args []pyObject) pyObject {
+	short := args[0].IsTruthy()
+
+	cmdIn := make([]pyObject, 3, 5)
+	cmdIn[0] = pyString("git")
+	cmdIn[1] = pyString("symbolic-ref")
+	cmdIn[2] = pyString("-q")
+	if short {
+		cmdIn = append(cmdIn, pyString("--short"))
+	}
+	cmdIn = append(cmdIn, pyString("HEAD"))
+
+	wantStdout := True
+	wantStderr := False
+	cacheOutput := True
+	keyExtra := pyString("")
+
+	execArgs := []pyObject{
+		pyList(cmdIn),
+		wantStdout,
+		wantStderr,
+		cacheOutput,
+		keyExtra,
+	}
+
+	return doExec(s, execArgs)
+}
+
+// execGitCommit returns the output of a git_commit() command.
+//
+// git_commit() returns the output of `git rev-parse --short HEAD`
+func execGitCommit(s *scope, args []pyObject) pyObject {
+	short := args[0].IsTruthy()
+
+	cmdIn := make([]pyObject, 2, 4)
+	cmdIn[0] = pyString("git")
+	cmdIn[1] = pyString("rev-parse")
+	if short {
+		cmdIn = append(cmdIn, pyString("--short"))
```

Comment:
The length of `--short` depends on your git config value of `core.abbrev`. This is a reproducibility issue.

I would recommend we not support hash truncation at this stage at all. I cannot think of any case where it would not be desirable from a collision-resistance perspective to always pass the full SHA. If someone only wants to actually display part of it in a `--version` or whatever, then they can truncate when printing.

Reply:
> The length of `--short` depends on your git config value of `core.abbrev`. This is a reproducibility issue.

Drat, I forgot about that configuration parameter.  Allowing anything that allows locale to slip through is also problematic, too (e.g. timezones).  I'm going to trim the amount of data that's included in format as a result and will dial back to a more austere set of format options.

> I would recommend we not support hash truncation at this stage at all. I cannot think of any case where it would not be desirable from a collision-resistance perspective to always pass the full SHA. If someone only wants to actually display part of it in a `--version` or whatever, then they can truncate when printing.

I'm unconvinced by any of the arguments made, but am going to take the position that there should be only one way to do things and that reproducibility is paramount (I don't see how this sha in this context could reliably used, but I digress).  This does shift the burden for formatting to the user, which I'm not wild about, but it leaves little ambiguity or chance for error due to user-specified configuration parameters in user-local git configs.  If someone does want the short SHA, they can specify `git_commit()[0:8]` and this value value will be stored in their `BUILD` files.